stan/math/rev/functor/adj_jac_apply.hpp

```diff
@@ -118,381 +48,389 @@ struct compute_dims<Eigen::Matrix<T, R, C>> {
+   * container for reverse pass.
+   */
+  std::array<size_t, 2> dims_;
+  /**
+   * Return the adjoint
+   * @tparam RetType Not called by user. Default of the return type of
+   * operator() of the functor F allows for SFINAE to choose the correct
+   * specialization given the required output type.
+   */
+  template <typename RetType = FReturnType,
+            require_arithmetic_t<RetType>* = nullptr>
+  inline auto& y_adj() const {
+    return y_vi_->adj_;
   }
 
-  template <typename... Pargs>
-  inline size_t count_memory(size_t count, const std::vector<var>& x,
-                             const Pargs&... args) {
-    static constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
-    offsets_[t] = count;
-    count += x.size();
-    return count_memory(count, args...);
+  /**
+   * Return a standard vector of the adjoints
+   * @tparam RetType Not called by user. Default of the return type of
+   * operator() of the functor F allows for SFINAE to choose the correct
+   * specialization given the required output type.
+   */
+  template <typename RetType = FReturnType,
+            require_std_vector_t<RetType>* = nullptr>
+  inline auto y_adj() const {
+    std::vector<double> var_y(dims_[0]);
+    for (size_t i = 0; i < dims_[0]; ++i) {
+      var_y[i] = y_vi_[i]->adj_;
+    }
+    return var_y;
   }
 
-  template <typename... Pargs>
-  inline size_t count_memory(size_t count, const std::vector<double>& x,
-                             const Pargs&... args) {
-    static constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
-    offsets_[t] = count;
-    return count_memory(count, args...);
+  /**
+   * Return an eigen matrix of the adjoints
+   * @tparam RetType Not called by user. Default of the return type of
+   * operator() of the functor F allows for SFINAE to choose the correct
+   * specialization given the required output type.
+   */
+  template <typename RetType = FReturnType, require_eigen_t<RetType>* = nullptr>
+  inline auto y_adj() const {
+    Eigen::Matrix<double, FReturnType::RowsAtCompileTime,
+                  FReturnType::ColsAtCompileTime>
+        var_y(dims_[0], dims_[1]);
+    const size_t iter_size{dims_[0] * dims_[1]};
+    for (size_t i = 0; i < iter_size; ++i) {
+      var_y(i) = y_vi_[i]->adj_;
+    }
+    return var_y;
   }
-
-  template <typename... Pargs>
-  inline size_t count_memory(size_t count, const std::vector<int>& x,
-                             const Pargs&... args) {
-    static constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
-    offsets_[t] = count;
-    return count_memory(count, args...);
+  /**
+   * Copy the vari memory from the input argument
+   * @tparam EigMat A type inheriting from `EigenBase` with a scalar type of
+   * vari.
+   * @param x An eigen input argument
+   */
+  template <typename EigMat, require_eigen_vt<is_var, EigMat>* = nullptr>
+  inline auto prepare_x_vis_impl(EigMat&& x) const {
+    ref_type_t<EigMat> x_ref(std::forward<EigMat>(x));
```

Comment:
The benefit is that it avoids a copy if the input doesn't need to be evaluated. Assigning to an ```Eigen::Ref``` will evaluate complex expressions, but only return a reference for trivial expressions (more background [here](https://eigen.tuxfamily.org/dox/classEigen_1_1Ref.html)).

Reply:
@t4c1 there's some weird catches that `ref_type_t` catches that `const auto& x_eval = x.eval()` does not right?

Reply:
`const auto&` should avoid the copying. The docs for `.eval()` for an Eigen::Matrix type are:

"Notice that in the case of a plain matrix or vector (not an expression) this function just returns a const reference, in order to avoid a useless copy." ([here](https://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#aa73e57a2f0f7cfcb4ad4d55ea0b6414b))

So unless there's something else I think `const auto&` is easier.

Reply:
Oh, a const ref would fail for rvalue types. Checkout the code for ref_type_t

Reply:
There's also a benefit for things like ```matrix.diagonal()``` or ```vector.transpose()```, where the ```.eval()``` would create a copy and the ```Eigen::Ref``` wouldn't

Reply:
Oh interesting. Alright refs are good.

Reply:
> There's also a benefit for things like matrix.diagonal() or vector.transpose(), where the .eval() would create a copy and the Eigen::Ref wouldn't

Not true (with default template params of `Eigen::Ref`)! 

By default `Ref` requires contiguous inner dimmension (each column must be contiguous for a col-major type), so `matrix.diagonal()` or `vector.transpose()` will make a copy. The same is true for `ref_type_t`. For `Ref` this requirement can be relaxed by specifying `StrideType = Eigen::Stride<>`. See https://eigen.tuxfamily.org/dox/classEigen_1_1Ref.html for details.